### PR TITLE
Add macOS

### LINKED ISSUE #12: Possible to create Mac version?
- [x] I read [the conda-forge documentation](https://conda-forge.org/docs/user/introduction.html#how-can-i-install-packages-from-conda-forge) and could not find the solution for my problem there.

Issue:

Hi, I was wondering if it might be possible to re-render this package and include a Mac version? I installed it via pip and that worked so I think it should be doable.

If there is interest, I can also try to create the PR but I just wanted to check first.

<br/>
Environment (<code>conda list</code>):
<details>

```
$ conda list

```
</details>

<br/>
Details about  <code>conda</code> and system ( <code>conda info</code> ):
<details>

```
$ conda info

```
</details>

### PR DESCRIPTION
**Pull request by BastianZim**
Checklist
* [x] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [ ] Bumped the build number (if the version is unchanged)
* [x] Reset the build number to `0` (if the version changed)
* [x] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [x] Ensured the license file is being packaged.

Closes #12 

TODO: 

* [x] Add back `gperftools`.
* [x] Fix Windows
* [x] Add tests

### CODE REVIEW

**Comment by BastianZim** [comment]
@conda-forge-admin, please rerender

**Comment by BastianZim** [comment]
@conda-forge-admin, please rerender

**Comment by BastianZim** [comment]
@conda-forge-admin, please rerender

**Comment by BastianZim** [comment]
@h-vetinari Just to keep you in the loop: I managed to build sentencepiece for linux and macOS by using PYPI and Grayskull but now Windows is crashing. 
I'll try some things but if everything fails I guess the easiest way would be to use PyPI as a source for macOS and keep everything else like before.

**Comment by BastianZim** [comment]
Here is someone with a similar problem: https://github.com/google/sentencepiece/issues/325

**Comment by BastianZim** [comment]
Also, I haven't figured out yet how to trigger the tests. Any help there would be appreciated.

**Comment by BastianZim** [comment]
@setu4993 Just to keep you in the loop as well: I managed to build sentencepiece for macOS by using pypi as a host. Unfortunately, this makes windows fail so I'm now going to split up the sources. Since linux works from both, the question would therefore be, if you'd prefer keeping linux on GitHub or if it should be moved to pypi?

**Comment by BastianZim** [comment]
Also, `gperftools` was in `build`, `host` and `run`. Is that on purpose? I'd assume that only `build` should be fine but just wanted to double check.

**Comment by BastianZim** [comment]
@conda-forge-admin, please rerender

**Comment by BastianZim** [comment]
@conda-forge-admin, please rerender

**Comment by BastianZim** [comment]
Also, just to add: Once this is merged it should be possible to try add ARM support using the bot.

**Comment by BastianZim** [comment]
@conda-forge-admin, please rerender

**Comment by BastianZim** [comment]
Seems like the windows problem is not related to PyPI but to pip. If anyone has an idea why let me know! 😄

**Comment by BastianZim** [comment]
@conda-forge-admin, please rerender

**Comment by BastianZim** [comment]
@conda-forge-admin, please rerender

**Comment by BastianZim** [comment]
@conda-forge-admin, please rerender

**Comment by BastianZim** [comment]
@setu4993 @h-vetinari Would be ready now.
Couple of comments (Was mentioned above but just as a summary):

1. I switched Linux and mac to PyPI as a source, this is recommended by conda-forge.
2. I removed `gperftools` from `host `and `run`. Didn't seem necessary. Let me know if this is needed.
3. Windows has to stay on GitHub, due to the error shown in https://github.com/conda-forge/sentencepiece-feedstock/runs/1844854080 which is caused by pip. I mentioned it in sentencepiece at https://github.com/google/sentencepiece/issues/591#issuecomment-774501757 but not sure if it gets fixed. Doesn't seem to be a huge problem though since conda supports multiple sources.
4. I removed the CLI tests since they're not working.
5. I updated it to 0.1.95.
6. I haven't added the tests from sentencepiece yet because I don't know how 😄 If anyone knows how, please let me know. Sentencepiece does have tests (https://github.com/google/sentencepiece/tree/master/python/test) that can be triggered somehow (https://github.com/google/sentencepiece/blob/ba7e11a17f606327d0652528d58d2dd8cd265c6f/python/setup.py#L131) but I don't know how so any help welcome.
7. Once this is merged, macOS ARM builds could be added via the bot: https://conda-forge.org/blog/posts/2020-10-29-macos-arm64/ I don't see a reason for this to not work so I'd add it, but I'd suggest doing it after the tests have been added to catch anything that a simple import might miss since the whole package is a bit finicky.

File: recipe/meta.yaml [hunk 1]
```diff
@@ -1,39 +1,41 @@
...
     - python
-    - gperftools  # [unix]
   run:
     - python
-    - gperftools  # [unix]
 
 test:
   imports:
     - sentencepiece
   commands:
```
Comment (h-vetinari) [inline recipe/meta.yaml:35]: Running the upstream test suite (if it is in the sources) can be as easy as:
```suggestion
  requires:
    - pytest
  source_files:
    - tests
  commands:
    - pytest tests
```

**Review by h-vetinari — CHANGES_REQUESTED** [review]
Hey @BastianZim 

Good to see you working on this. You need to be aware of the limitations of `grayskull` (or `conda skeleton`) in modifying existing recipes.

* We cannot disable the CLI-tests. If they are broken, the package is likely broken.
* We should really _really_ run the upstream test suite. This is easy to do when using the github sources (unless the tests are also in the PyPI package, which is often not the case).
* I disagree that building from PyPI is preferred - building from source might be harder, but gives a much clearer picture of what is being done (which is often one of the key contributions of the conda-forge ecosystem to make sure it's done correctly, or at least consistently with the rest of the ecosystem).
* I really dislike using separate sources.

**Comment by BastianZim** [comment]
Hi @h-vetinari thanks for the feedback! My comments:

> * We cannot disable the CLI-tests. If they are broken, the package is likely broken.

I think the CLI is not included when you're building from PyPI with pip. I verified this by installing the package in a test environment using pip and no CLI was available, it's only there when manually building the package. Therefore I think it should be removed when installing it using PyPI and pip.

> * We should really really run the upstream test suite. This is easy to do when using the github sources (unless the tests are also in the PyPI package, which is often not the case).

Just added it, let's see if it works. The tests are included with PyPi. As I said, the reason for not including it was just that I don't know how. 😄 

> * I disagree that building from PyPI is preferred - building from source might be harder, but gives a much clearer picture of what is being done (which is often one of the key contributions of the conda-forge ecosystem to make sure it's done correctly, or at least consistently with the rest of the ecosystem).

I just followed the recommendations in the example meta.yaml, there PyPI is the first recommendation. From what I can see in the package downloaded from PyPI, it actually still builds it the exact same way using the script that was previously in `build.sh`, with the only difference being that it works. This is the script included with the PyPI download: https://github.com/google/sentencepiece/blob/master/python/build_bundled.sh

> * I really dislike using separate sources.
Yep, agree but I just can't make it work otherwise. 😄

**Comment by BastianZim** [comment]
Just to add because I forgot it above:

> You need to be aware of the limitations of grayskull (or conda skeleton) in modifying existing recipes.

Yep, fully agree. This is actually heavily modified from Grayskull. The original meta.yaml is:

```yaml
{% set name = "sentencepiece" %}
{% set version = "0.1.95" %}


package:
  name: {{ name|lower }}
  version: {{ version }}

source:
  url: https://pypi.io/packages/source/{{ name[0] }}/{{ name }}/sentencepiece-{{ version }}.tar.gz
  sha256: 8dd6e3e110f4c3f85a46e4a2ae1b6f7cf020b907fab50eac22beccf1680e0ea5

build:
  number: 0
  noarch: python
  script: {{ PYTHON }} -m pip install . -vv

requirements:
  host:
    - pip
    - python
  run:
    - python

test:
  imports:
    - sentencepiece/__init__
    - sentencepiece/sentencepiece_model_pb2
  commands:
    - pip check
  requires:
    - pip

about:
  home: https://github.com/google/sentencepiece
  summary: SentencePiece python wrapper
  license: Apache-2.0
  license_file: LICENSE

extra:
  recipe-maintainers:
    - ADD_YOUR_GITHUB_ID_HERE
```

I just used it as a starting point.

**Comment by BastianZim** [comment]
@conda-forge-admin, please rerender

**Comment by BastianZim** [comment]
Moving this back to draft until the tests have been added.

**Comment by BastianZim** [comment]
@conda-forge-admin, please rerender

**Comment by BastianZim** [comment]
@h-vetinari Answered your comments in https://github.com/conda-forge/sentencepiece-feedstock/pull/17#issuecomment-774518350
Just one open questions I'm adding here: I'm now installing it with pip as outlined in the linked reply. What's your opinion on that? It seems to be preferred but I just wanted to get your input. If you agree I'd try switching win over to it as well although I'm sceptical since it failed with the PyPI source before.

**Comment by BastianZim** [comment]
@conda-forge-admin, please rerender

**Comment by BastianZim** [comment]
@conda-forge-admin, please rerender

**Comment by BastianZim** [comment]
Another update from my side @h-vetinari @setu4993 

* I switched Linux back to the original build way. CLI tests are included again and passing, same for upstream tests.
* macOS is now being built from GitHub using pip. This means, that the CLI is not available but that is the same for the Windows build. I tried building macOS using a custom script, which works (#18), but the CLI is still not included there, no idea why. For simplicity, I, therefore, stayed with pip. Might also be easier to add mac-arm support that way but not sure about that.
* `gperftools` is still only in build. Let me know if it should be added back.

This PR would thus be ready from my side since I cannot figure out a way to build the CLI and whilst it's nice to have, I think having macOS at all is already helpful so I would suggest merging this and then someone else can take a shot at, if they need the CLI.

I added quite a lot of unnecessary commits now so I'd be happy to create a separate PR and only include the necessary ones. @setu4993 Let me know what you prefer.

**Comment by setu4993** [comment]
Thanks for your work on this, @BastianZim. I've been passively following along the progress on this so far, and appreciate it.

> macOS is now being built from GitHub using pip.

My preference would be to build from source directly, not with `pip`, as you have done in #18 (glad that you got it working, I struggled with it for weeks!). That would also make it equivalent to all other distributions since Linux and Windows are both built from source.

Skipping CLI tests for MacOS and Windows seems fine to me.

> `gperftools` is still only in build. Let me know if it should be added back.

I think that's correct.

> I added quite a lot of unnecessary commits now so I'd be happy to create a separate PR and only include the necessary ones.

No worries on that. We can squash and merge the PR at the end to keep the tree on `master` clean. Do what works best for you :).

**Comment by BastianZim** [comment]
Thanks for your feedback @setu4993! My comments below:

RE pip: I'll change that, no problem, just the following:

* pip actually builds from source as well and runs the exact same script (My implementation is the same as theirs, see: https://github.com/google/sentencepiece/blob/master/python/build_bundled.sh) so there isn't any real difference but keeping everything synchronous makes sense.
* The build section (https://github.com/google/sentencepiece#build-and-install-sentencepiece-command-line-tools-from-c-source) lists the following for macOS: `sudo update_dyld_shared_cache` which will conflict at some point with the planned removal of sudo on mac (https://github.com/conda-forge/conda-forge.github.io/issues/810). I, therefore, excluded it which still passed the test but that is something that should be kept in mind in case that ever fails.

RE PR: I'll create a new one once we have agreed on everything. Seems cleaner and easier to follow in case something has to be changed later.

**Comment by BastianZim** [comment]
Closing for #19

**Pull request closed without merging**